### PR TITLE
Make AppName configurable from within ConfigLoader

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
Fixes https://github.com/awslabs/smithy-rs/issues/2511

## Description
This small PR adds a setter for `app_name` in `ConfigLoader`. Prior to the commit, in a typical config chain of `ConfigLoader` -> `SdkConfig` -> service config, the field was only settable from `SdkConfig` onward.

## Testing
- [x] Added unit test
- [x] Passed tests in CI  

## Checklist
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: CHANGELOG.next.toml [hunk 1]
```diff
@@ -52,3 +52,9 @@ message = "The `enableNewCrateOrganizationScheme` codegen flag has been removed.
 references = ["smithy-rs#2507"]
 meta = { "breaking" = false, "tada" = false, "bug" = false, "target" = "client" }
 author = "jdisanti"
+
+[[aws-sdk-rust]]
+message = "Make AppName configurable from within `ConfigLoader`."
```
Comment (jdisanti) [inline CHANGELOG.next.toml:57]:
```suggestion
message = "`AppName` is now configurable from within `ConfigLoader`."
```

File: CHANGELOG.next.toml [hunk 2]
```diff
@@ -52,3 +52,9 @@ message = "The `enableNewCrateOrganizationScheme` codegen flag has been removed.
...
+references = ["smithy-rs#2513"]
+meta = { "breaking" = false, "tada" = false, "bug" = false }
```
Comment (jdisanti) [inline CHANGELOG.next.toml:59]:
```suggestion
meta = { "breaking" = false, "tada" = false, "bug" = true }
```

File: aws/rust-runtime/aws-config/src/lib.rs [hunk 3]
```diff
@@ -344,6 +344,24 @@ mod loader {
             self
         }
 
+        /// Override the name of the app used to build [`SdkConfig`](aws_types::SdkConfig).
```
Comment (jdisanti) [inline aws/rust-runtime/aws-config/src/lib.rs:347]:
```suggestion
        /// Override the name of the app used to build [`SdkConfig`](aws_types::SdkConfig).
        ///
```

**Review by jdisanti — APPROVED** [review]

File: aws/rust-runtime/aws-config/src/lib.rs [hunk 4]
```diff
@@ -344,6 +344,25 @@ mod loader {
...
+        ///
+        /// This _optional_ name is used to identify the application in the user agent that
+        /// gets sent along with requests.
```
Comment (pose) [inline aws/rust-runtime/aws-config/src/lib.rs:350]: Why is it call `AppName` and not `UserAgent`?

File: aws/rust-runtime/aws-config/src/lib.rs [hunk 5]
```diff
@@ -344,6 +344,25 @@ mod loader {
...
+        /// ```no_run
+        /// # async fn create_config() {
+        /// use aws_config::AppName;
+        /// let config = aws_config::from_env()
+        ///     .app_name(AppName::new("my-app-name").expect("valid app name"))
+        ///     .load().await;
+        /// # }
+        /// ```
+        pub fn app_name(mut self, app_name: AppName) -> Self {
+            self.app_name = Some(app_name);
```
Comment (pose) [inline aws/rust-runtime/aws-config/src/lib.rs:362]: Do we need to add any sort of validation for this being non-empty string or is that handled by `AppName`?

**Review by pose — CHANGES_REQUESTED** [review]

File: aws/rust-runtime/aws-config/src/lib.rs (hunk 4 above)
Comment (rcoh) [reply aws/rust-runtime/aws-config/src/lib.rs:350]: it's not the `UserAgent` — it eventually gets composed as part of the user agent. But the user agent itself has far more in it than the app name (platform, version, etc.)

**Review by rcoh — APPROVED** [review]

File: aws/rust-runtime/aws-config/src/lib.rs (hunk 5 above)
Comment (ysaito1001) [reply aws/rust-runtime/aws-config/src/lib.rs:362]: Thanks for pointing it out! It's good to guard against that. Addressed in b296de9.

**Review by pose — APPROVED** [review]

**Pull request merged**